flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecMultiJoin.java

```diff
@@ -0,0 +1,444 @@
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Stream {@link StreamExecNode} for N-way Joins. This node handles multi-way joins in streaming
+ * mode, supporting different join types and conditions for each input.
+ */
+@ExecNodeMetadata(
+        name = "stream-exec-multi-join",
+        version = 1,
+        producedTransformations = StreamExecMultiJoin.MULTI_JOIN_TRANSFORMATION,
+        minPlanVersion = FlinkVersion.v2_0,
+        minStateVersion = FlinkVersion.v2_0)
+public class StreamExecMultiJoin extends ExecNodeBase<RowData>
+        implements StreamExecNode<RowData>, SingleTransformationTranslator<RowData> {
+
+    public static final String MULTI_JOIN_TRANSFORMATION = "multi-join";
+
+    // Field names for JSON serialization
+    private static final String FIELD_NAME_ID = "id";
+    private static final String FIELD_NAME_TYPE = "type";
+    private static final String FIELD_NAME_CONFIGURATION = "configuration";
+    private static final String FIELD_NAME_INPUT_PROPERTIES = "inputProperties";
+    private static final String FIELD_NAME_OUTPUT_TYPE = "outputType";
+    private static final String FIELD_NAME_DESCRIPTION = "description";
+    private static final String FIELD_NAME_JOIN_TYPES = "joinTypes";
+    private static final String FIELD_NAME_JOIN_CONDITIONS = "joinConditions";
+    private static final String FIELD_NAME_GENERATED_MULTI_JOIN_CONDITION =
+            "generatedMultiJoinCondition";
+    private static final String FIELD_NAME_JOIN_ATTRIBUTE_MAP = "joinAttributeMap";
+    private static final String FIELD_NAME_INPUT_SIDE_SPECS = "inputSideSpecs";
+    private static final String FIELD_NAME_STATE_METADATA_LIST = "stateMetadataList";
+    private static final String FIELD_NAME_KEY_SELECTORS = "keySelectorsForTransformation";
+    private static final String FIELD_NAME_COMMON_KEY_TYPE = "commonKeyType";
+
+    @JsonProperty(FIELD_NAME_JOIN_TYPES)
+    private final List<JoinType> joinTypes;
+
+    @JsonProperty(FIELD_NAME_JOIN_CONDITIONS)
+    private final List<? extends @Nullable RexNode> joinConditions;
+
+    @JsonProperty(FIELD_NAME_JOIN_ATTRIBUTE_MAP)
+    private final Map<
+                    Integer,
+                    Map<
+                            AttributeBasedJoinKeyExtractor.AttributeRef,
+                            AttributeBasedJoinKeyExtractor.AttributeRef>>
+            joinAttributeMap;
+
+    @JsonProperty(FIELD_NAME_INPUT_SIDE_SPECS)
+    private final List<JoinInputSideSpec> inputSideSpecs;
+
+    @JsonProperty(FIELD_NAME_KEY_SELECTORS)
+    private final List<KeySelector<RowData, RowData>> commonJoinKeySelectorsForTransformation;
```

Comment:
Having a KeySelector here is very uncommon. Usually, the KeySelector should be created within the `translateToPlan` method. Ideally this field uses easy to serialize arrays or list/map. You will spot why when you implement a `RestoreTest` for it which tests the compiled plan serde.

Reply:
Ah, interesting. Thanks for the hint. I passed it down because we had all information to create it in the logical level but I can create it here.

I need the joinAttributeMap for that which also has a custom `AttributeRef` class, however this class already implements Serializable. So I suppose that's ok?

Reply:
Implementing `Serializable` is not enough. The CompiledPlan is designed for a long-term, Java-independent representation, so you will need to implement custom serializers/deserialized if required. Of course this should be avoided for keeping the serde stack of CompiledPlan simple. Maybe thing about how you can represent it as standard collections? Otherwise a custom serializer/deserializer is an option, not many operators do this, but some.